composer/cli/launcher.py

```diff
@@ -0,0 +1,201 @@
+                    else:
+                        output = process.stdout.read()
+
+                    if process.stderr is None:
+                        stderr = ""
+                    else:
+                        stderr = process.stderr.read()
+                    exc = subprocess.CalledProcessError(
+                        process.returncode,
+                        cmd=process.args,
+                        output=output,
+                        stderr=stderr,
+                    )
+                    error_msg = [
+                        "Error in subprocess",
+                        "----------Subprocess STDOUT----------",
+                        exc.output,
+                        "----------Subprocess STDERR----------",
+                        exc.stderr,
+                    ]
+                    print("\n".join(error_msg))
+                    print(exc)
+                    sys.exit(process.returncode)
+                else:
+                    # exited cleanly
+                    processes.remove(process)
+                    break
+        time.sleep(1)
+
+
+def cleanup_processes(processes: Set[subprocess.Popen]):
+    if len(processes) == 0:
+        if torch.distributed.is_initialized():
+            torch.distributed.destroy_process_group()
+        return
+
+    for process in processes:
+        if process.returncode is None:
+            print(f"Killing subprocess {process.pid} with SIGTERM")
+            try:
+                os.killpg(process.pid, signal.SIGTERM)
+            except ProcessLookupError:
+                pass
+
+    current_time = datetime.datetime.now()
+    print(f"Waiting {CLEANUP_TIMEOUT.seconds} seconds for processes to terminate...")
+    while datetime.datetime.now() - current_time < CLEANUP_TIMEOUT:
+        if all(process.returncode is not None for process in processes):
+            break
+        time.sleep(0.1)
+
+    for process in processes:
+        if process.returncode is None:
+            print(f"Failed to kill subprocess {process.pid} with SIGTERM; using SIGKILL instead")
+            try:
+                os.killpg(process.pid, signal.SIGKILL)
+            except ProcessLookupError:
+                pass
+
+    if torch.distributed.is_initialized():
+        torch.distributed.destroy_process_group()
+
+
+def main():
+    args = parse_args()
+
+    processes = launch_processes(nproc=args.nproc,
+                                 world_size=args.world_size,
+                                 base_rank=args.base_rank,
+                                 master_addr=args.master_addr,
+                                 master_port=args.master_port,
+                                 module_mode=args.module_mode,
+                                 training_script=args.training_script,
+                                 training_script_args=args.training_script_args)
+
+    try:
+        monitor_processes(processes)
+    finally:
+        cleanup_processes(processes)
+        return 0
```

Comment:
If there's an error, let's return an error code rather than 0. Perhaps the code of the first process that returns a nonzero code.